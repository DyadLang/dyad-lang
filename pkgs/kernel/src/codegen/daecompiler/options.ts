export interface DAECodeGenerationOptions {
}

export function normalizeDAECGOptions(
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  opts: Partial<DAECodeGenerationOptions>
): DAECodeGenerationOptions {
  return {
  };
}
