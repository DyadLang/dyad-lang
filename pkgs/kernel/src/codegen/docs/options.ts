import { assertUnreachable } from "@juliacomputing/dyad-common";
import {
  getDefinitionEntity,
  QueryHandler,
  ResolvedType,
  unparseDefinitionEntity,
} from "../../workspace/index.js";
import { codeFence } from "./primitives.js";
import { Definition } from "@juliacomputing/dyad-ast";
import { Just, Maybe, Nothing } from "purify-ts/Maybe";

export interface RenderOptions {
  /**
   * Generate href to another entity.  If no options is provided here, the URN
   * is used (which is unlikely to be useful, so this will almost always be
   * overridden).
   *
   * @param entity
   * @returns URL to the specified entity
   */
  generateHref?: (query: QueryHandler, type: ResolvedType) => Maybe<string>;
  /**
   * Generate HTML within Markdown.  If this isn't specified, then the HTML is
   * just written inline (which should normally be the case for Markdown).  In
   * the case of Documenter.jl, this isn't allowed and a special code fence
   * needs to be used.
   *
   * @param html
   * @returns
   */
  html?: (html: string) => string;
}

/**
 * A version of `RenderOptions` where all optional fields have been
 * filled in either be defaults or overrides.
 */
export type NormalizedRenderOptions = Required<RenderOptions>;

/**
 * Normalize an instance of `RenderOptions`
 *
 * @param opts Raw `RenderOptions`
 * @returns `NormalizedRenderOptions`
 */
export function normalizeRenderOptions(
  opts: RenderOptions
): NormalizedRenderOptions {
  const defaultHref = (): Maybe<string> => Nothing;
  const defaultHtml = (html: string): string => html;
  return {
    generateHref: opts.generateHref ?? defaultHref,
    html: opts.html ?? defaultHtml,
  };
}

/**
 * These are the options that should be used in the cae of documentations
 * generated by Documenter.jl.
 */
export const documenterOptions: NormalizedRenderOptions = {
  generateHref: (e, t) =>
    standardDocumenterHrefs(
      process.env["DYAD_STD_BASE"] ??
        "https://vigilant-adventure-ggzv8m1.pages.github.io/dev/stdlibs",
      process.env["DYAD_DOCS_BASE"] ?? "/",
      e,
      t
    ),
  html: (html: string) => codeFence("@raw html", html),
};

const standardLibraries: string[] = [
  "Dyad",
  "BlockComponentS",
  "ElectricalComponents",
  "TranslationalComponents",
  "RotationalComponents",
  "ThermalComponents",
  "HydraulicComponents",
];

function standardDocumenterHrefs(
  stdBaseURL: string,
  docsBaseURL: string,
  query: QueryHandler,
  rt: ResolvedType
): Maybe<string> {
  const baseOfDef = (def: Definition) => {
    const entity = query(getDefinitionEntity(def));
    const { library, modules } = unparseDefinitionEntity(entity);
    const isStandard = library in standardLibraries;
    const modpath = modules.length === 0 ? "" : modules.join("/");
    const base = isStandard
      ? `${stdBaseURL}/${modpath}`
      : `${docsBaseURL}/${modpath}`;
    return base;
  };
  if (rt.resolves === "scalar") {
    return rt.derived.caseOf({
      Nothing: () => Nothing,
      Just: (d) => Just(`${baseOfDef(d)}/types/${d.name.value}`),
    });
  }
  const base = baseOfDef(rt.def);
  switch (rt.resolves) {
    case "struct":
    case "fun":
    case "enum":
      return Just(`${base}/types/${rt.def.name.value}`);
    case "adef":
      return Just(`${base}/analyses/${rt.def.name.value}`);
    case "cdef": {
      const qual = rt.def.qualifier;
      switch (qual) {
        case "partial":
          return Just(`${base}/types/${rt.def.name.value}`);
        case "test":
          return Just(`${base}/tests/${rt.def.name.value}`);
        case "external":
        case "example":
        case null:
          return Just(`${base}/components/${rt.def.name.value}`);
        default:
          assertUnreachable(qual);
      }
    }
    case "sclcon":
      return Just(`${base}/connectors/${rt.def.name.value}`);

    case "strcon":
      return Just(`${base}/connectors/${rt.def.name.value}`);
    default:
      assertUnreachable(rt);
  }
}
