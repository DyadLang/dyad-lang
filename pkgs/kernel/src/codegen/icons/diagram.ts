import { QueryHandler } from "../../workspace/index.js";
import { diagramLayout } from "./layout.js";
import {
  renderComponent,
  renderConnection,
  renderDiagram,
  renderJunction,
  renderLabel,
} from "./render.js";
import { ModelInstance } from "../../instantiate/model.js";
import { toJS } from "mobx";
import { Result } from "@juliacomputing/dyad-common";

/**
 * The purpose of this function is to generate the icon "markup" for a given
 * model definition.  Currently, the assumption is that this "markup" is in
 * SVG although something like web components could be used here as well since
 * the assumption is that this will be injected into the DOM.
 *
 * @param def Definition whose icon is being requested
 * @param iconName The name of the icon to render (a given definition may have multiple)
 * @param workspace The workspace that this definition is managed in
 * @param visited The set of all model definitions already called (to avoid infinite recursion)
 * @returns
 */
export async function generateDiagram(
  instanceName: string,
  instance: ModelInstance,
  query: QueryHandler
): Promise<Result<string>> {
  const _def = toJS(query(instance.def));
  // Generate the layout of the components, connections and connectors.  This is just
  // information about the placement of all the icons and lines.
  return (await diagramLayout(instance, query)).map((layout) => {
    // Render the components
    const components = Object.values(layout.components).map((g) =>
      renderComponent(g, { filter: "url(#drop-shadow)" })
    );

    // Render the connections
    const connections = Object.values(layout.connections).map((g) =>
      renderConnection(g, g.pathAttrs)
    );

    // Render the connectors
    const connectors = Object.values(layout.connectors).map((g) =>
      renderComponent(g, {})
    );

    const junctions = layout.junctions.map((g) => renderJunction(g, {}));

    const labels = layout.labels.map((g) => renderLabel(g, 0));

    // Now wrap it all in an SVG wrapper
    return renderDiagram(
      instanceName,
      components,
      connections,
      connectors,
      junctions,
      labels
    );
  });
}
