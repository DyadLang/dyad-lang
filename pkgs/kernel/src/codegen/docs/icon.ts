import { Lines } from "@juliacomputing/dyad-common";
import { NormalizedRenderOptions } from "./options.js";

export function renderIcon(
  lines: Lines,
  options: NormalizedRenderOptions,
  icon: string,
  iconName: string
) {
  lines.add(
    options.html(`
      <div style="float: right;width: 20%;max-width: 10vw; margin-left: 20px">
      ${icon}
      <span style="width: 100%; text-align: center; display:inline-block"><code>${iconName}</code> Icon</span>
      </div>
    `)
  );
}
