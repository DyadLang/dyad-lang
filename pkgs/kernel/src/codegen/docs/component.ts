import {
  Lines,
  partialResult,
  PartialResult,
  Problem,
  SuccessfulResult,
} from "@juliacomputing/dyad-common";
import { codeFence, dyadCodeFenceLanguage, inlineCode } from "./primitives.js";
import { ModelInstance } from "../../instantiate/model.js";
import { unparseExpression, unparseDyad } from "@juliacomputing/dyad-parser";
import {
  connectorDocString,
  parameterDocString,
  variableDocString,
} from "../mtk/docs.js";
import { ComponentDefinition } from "@juliacomputing/dyad-ast";
import { NormalizedRenderOptions } from "./options.js";
import { renderIcon } from "./icon.js";
import { QueryHandler } from "../../workspace/selector.js";
import { resolveType } from "../../workspace/index.js";

export interface ComponentSlots {
  pre_source?: string;
  pre_related?: string;
}

export function renderComponentMarkdown(
  options: NormalizedRenderOptions,
  instance: ModelInstance,
  def: ComponentDefinition,
  flattened: ComponentDefinition,
  tests: string[],
  slots: ComponentSlots,
  query: QueryHandler,
  icon: string,
  diagram?: string
): SuccessfulResult<string> | PartialResult<string> {
  const lines = new Lines("");
  const problems: Problem[] = [];
  renderIcon(lines, options, icon, def.name.value);
  lines.add(`# ${inlineCode(def.name.value)}`);
  if (instance.doc_string !== null) {
    lines.add("");
    lines.add(instance.doc_string.value);
  }
  if (def.extends.length > 0) {
    lines.add("");
    for (const e of def.extends) {
      const rt = query(resolveType(e.name, def, []));
      rt.ifResult((t) => {
        const href = options.generateHref(query, t);
        const label = inlineCode(e.name.map((n) => n.value).join("."));
        lines.add(
          href.caseOf({
            Nothing: () => label,
            Just: (url) => `[${label}](${url})`,
          })
        );
      }, problems);
    }
  }
  lines.add("");
  lines.add("## Usage");
  lines.add("");
  lines.add(inlineCode(`${def.name.value}(${parameterList(instance)})`));
  lines.add("");
  lines.add(
    parameterDocString(
      Object.entries(instance.structural),
      Object.entries(instance.parameters).filter((x) => !x[1].final)
    )
  );
  lines.add("");
  lines.add(connectorDocString(Object.entries(instance.connectors)));
  lines.add("");
  lines.add(variableDocString(Object.entries(instance.variables)));

  if (slots.pre_source) {
    lines.add("");
    lines.add(slots.pre_source);
  }
  lines.add("");

  lines.add("## Source");
  lines.add(codeFence(dyadCodeFenceLanguage, unparseDyad(def)));
  lines.add("");
  lines.add(
    options.html(`
<details>
<summary>Flattened Source</summary>
`)
);
lines.add(codeFence(dyadCodeFenceLanguage, unparseDyad(flattened)));
lines.add(options.html(`</details>`));
  lines.add(
    options.html(`
<br></br>
`));
  if (diagram !== undefined && Object.entries(instance.components).length > 0) {
    lines.add("");
    lines.add("## Diagram");
    lines.add(
      options.html(`
      <div style="">
      ${diagram}
      </div>
    `)
    );
  }
  if (slots.pre_related) {
    lines.add("");
    lines.add(slots.pre_related);
  }
  lines.add("");
  lines.add("## Related");
  lines.add("");
  lines.add("- Examples");
  lines.add("- Experiments");
  lines.add("- Analyses");
  if (tests.length > 0) {
    lines.add("- Tests");
    for (const t of tests) {
      lines.add(`  - [${inlineCode(t)}](#)`);
    }
  }
  return partialResult(lines.toString(), ...problems);
}

function parameterList(instance: ModelInstance): string {
  return Object.entries(instance.parameters)
    .map(([k, v]) =>
      v.default.mapOrDefault((e) => `${k}=${unparseExpression(e)}`, `${k}`)
    )
    .join(", ");
}
