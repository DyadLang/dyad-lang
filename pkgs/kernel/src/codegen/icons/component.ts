import { Just, Maybe, Nothing } from "purify-ts/Maybe";
import { ComponentInstance } from "../../instantiate/component.js";
import { viewHeight, viewWidth } from "./constants.js";
import { generateIcon } from "./icon.js";
import { DyadWayPoint } from "../../metadata/connection.js";
import { Transformation } from "./transform.js";
import { ModelInstance } from "../../instantiate/model.js";
import { NormalizedDyadDefinition } from "../../metadata/index.js";
import { unparseMTKExpression } from "../equation.js";
import { Problem, Result, successfulResult } from "@juliacomputing/dyad-common";
import { ComponentDefinition, hasExpression } from "@juliacomputing/dyad-ast";
import { QueryHandler } from "../../workspace/selector.js";

export interface ComponentGraphic {
  componentName: string;
  transform: Transformation;
  icon: string;
}

export interface LabelGraphic {
  text: string;
  rot: number;
  location: DyadWayPoint;
  attrs: Record<string, string>;
}

export async function componentGraphics(
  instance: ModelInstance,
  query: QueryHandler
): Promise<Record<string, ComponentGraphic>> {
  const ret: Record<string, ComponentGraphic> = {};
  // Inject icons for all subcomponents
  // TODO: Optimize this so async calls are concurrent
  const problems: Problem[] = [];
  for (const [cn, comp] of Object.entries(instance.components)) {
    await comp().ifAsyncResult(async (compi) => {
      const def = query(compi.instance.def);
      (await componentGraphic(cn, compi, def, query)).map((graphic) => {
        graphic.ifJust((g) => (ret[cn] = g));
      });
    }, problems);
  }
  return ret;
}

export async function componentGraphic(
  componentName: string,
  comp: ComponentInstance,
  cdef: ComponentDefinition,
  query: QueryHandler
): Promise<Result<Maybe<ComponentGraphic>>> {
  const placement = comp.metadata.Dyad.placement;
  if (placement === undefined || placement.diagram === undefined) {
    return successfulResult(Nothing);
  }
  const diagram = placement.diagram;
  const subs: Record<string, string> = { instance: componentName };
  for (const [key, val] of Object.entries(comp.mods)) {
    if (hasExpression(val)) {
      subs[key] = unparseMTKExpression(val.expr);
    }
  }
  return (
    await generateIcon(subs, cdef, diagram.iconName, diagram.rot, query)
  ).map((icon) => {
    const x1 = diagram.x1;
    const y1 = diagram.y1;
    const x2 = diagram.x2;
    const y2 = diagram.y2;
    const w = x2 - x1;
    const sw = w / viewWidth;
    const h = y2 - y1;
    const sh = h / viewHeight;
    const rot = diagram.rot;
    const transform = { x1, x2, y1, y2, sw, sh, rot };

    return Just({
      componentName,
      transform,
      icon,
    });
  });
}

export function labelGraphics(
  substitutions: Record<string, string>,
  metadata: NormalizedDyadDefinition,
  layer: "icon" | "diagram"
): LabelGraphic[] {
  const ret: LabelGraphic[] = [];

  for (const label of metadata.Dyad.labels) {
    if (label.layer === layer || label.layer === "all") {
      let text = label.label;
      for (const [v, sub] of Object.entries(substitutions)) {
        text = text.replaceAll(`$(${v})`, sub);
      }

      // These are the default values, potentially overridden by the
      const attrs: Record<string, string> = {
        fill: "black",
        "font-size": "200",
        "font-weight": "200",
        "dominant-baseline": "central",
        "text-anchor": "middle",
      };
      ret.push({
        text,
        rot: label.rot,
        location: { x: label.x, y: label.y },
        attrs: { ...attrs, ...label.attrs },
      });
    }
  }
  return ret;
}
