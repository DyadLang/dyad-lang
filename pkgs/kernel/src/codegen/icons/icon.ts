import {
  Definition,
  FileLevelNode,
  MetadataNode,
  isComponentDefinition,
  isDefinition,
} from "@juliacomputing/dyad-ast";
import { problemSpan, QueryHandler } from "../../workspace/index.js";
import { instantiateModel } from "../../instantiate/index.js";
import {
  dyadDefinition,
  normalizeDefinition,
} from "../../metadata/definition.js";
import { connectorLayer } from "./layers.js";
import {
  Nullable,
  Problem,
  buf2str,
  createError,
  Result,
  partialResult,
  successfulResult,
} from "@juliacomputing/dyad-common";
import { zodIssue2Problem } from "../../metadata/issues.js";
import { defaultIcon } from "../../metadata/default_icons.js";
import { labelGraphics } from "./component.js";
import { renderLabel } from "./render.js";
import { unparseMTKExpression } from "../equation.js";

const IconNotFound = createError("icon-not-found", "Icon Not Found");
const UnsupportedContentType = createError(
  "unsupported-content-type",
  "Unsupported Content Type"
);
/**
 * The purpose of this function is to generate the icon "markup" for a given
 * model definition.  Currently, the assumption is that this "markup" is in
 * SVG although something like web components could be used here as well since
 * the assumption is that this will be injected into the DOM.
 *
 * @param def Definition whose icon is being requested
 * @param iconName The name of the icon to render (a given definition may have multiple)
 * @param workspace The workspace that this definition is managed in
 * @param visited The set of all model definitions already called (to avoid infinite recursion)
 * @returns
 */
export async function generateIcon(
  substitutions: Record<string, string>,
  def: Definition,
  iconName: string,
  baserot: number,
  query: QueryHandler
): Promise<Result<string>> {
  const entries: string[] = [];
  const subs = { ...substitutions };
  const problems: Array<Problem<unknown>> = [];

  // Inject this components icon as the "backdrop" for the icon
  (await getIcon(def.name.value, def, def.metadata, iconName, query)).map(
    (myIcon) =>
      entries.push(
        `<g transform="translate(0 0) scale(1.0) rotate(0)" transform-origin="center center">${myIcon}</g>`
      )
  );

  if (isComponentDefinition(def)) {
    const rp = instantiateModel(def, {}, null, query);
    const vr = await rp.mapAsync(async (instance) => {
      // Then inject the icon for all the connectors.
      // FIX: Use Promise.all()
      (await connectorLayer(instance, baserot, query)).ifResult((lines) => {
        entries.push(...lines);
      }, problems);

      for (const [pn, p] of Object.entries(instance.parameters)) {
        p.default.ifJust((e) => (subs[pn] = unparseMTKExpression(e)));
        if (subs[pn] === undefined) {
          subs[pn] = `$(${pn})`;
        }
      }
    });
    problems.push(...vr.problems());
  }

  if (isDefinition(def)) {
    const metadata = dyadDefinition
      .passthrough()
      .safeParse(def.metadata ? def.metadata.value : {});
    const norm = normalizeDefinition(metadata.success ? metadata.data : {});

    for (const label of labelGraphics(subs, norm, "icon")) {
      entries.push(renderLabel(label, baserot));
    }
  }

  // Now wrap it all in an SVG wrapper
  const ret = `<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 1000 1000"
    overflow="visible" shape-rendering="geometricPrecision" text-rendering="geometricPrecision">
    <style>
    text {
      transform-origin: center center;
    }
    </style>${entries.join("\n")}
    </svg>`;

  return partialResult(ret, ...problems);
}

const issueMap = zodIssue2Problem("getIcon", "definition");

export async function getIcon(
  context: string,
  parent: FileLevelNode,
  metadata: Nullable<MetadataNode>,
  iconName: string,
  query: QueryHandler
): Promise<Result<string>> {
  /** If there is no metadata at all, return default icon */
  if (metadata === null) {
    return successfulResult(defaultIcon);
  }

  /** Validate metadata against expected schema */
  const validated = dyadDefinition.safeParse(metadata ? metadata.value : {});
  if (!validated.success) {
    const problems = validated.error.issues.map((x) =>
      issueMap(x, problemSpan(parent, metadata.span))
    );

    return partialResult(defaultIcon, ...problems);
  }
  const defmeta = normalizeDefinition(validated.data);
  let iconUrl: string | undefined = defmeta.Dyad.icons[iconName];
  const problems: Problem[] = [];
  if (iconUrl === undefined) {
    problems.push(
      new IconNotFound(
        iconName,
        `Icon ${iconName} in ${context} not found in metadata`
      )
    );
    const entries = Object.entries(defmeta.Dyad.icons);
    if (entries.length === 0) {
      return partialResult(defaultIcon, ...problems);
    }
    iconUrl = entries[0][1];
  }
  if (iconUrl === undefined) {
    return partialResult(defaultIcon, ...problems);
  }

  const iconToUse = iconUrl;

  const contents = query(({ fetch }) => fetch(iconUrl));
  if (!contents.hasValue()) {
    return partialResult(defaultIcon, ...contents.problems());
  }
  const v = contents.value;
  if (v.contentType.startsWith("image/svg")) {
    return successfulResult(buf2str(v.contents));
  }
  return partialResult(
    defaultIcon,
    new UnsupportedContentType(
      iconToUse,
      `Unable to handle content type ${v.contentType} for ${iconUrl}, using default icon`
    )
  );
}
