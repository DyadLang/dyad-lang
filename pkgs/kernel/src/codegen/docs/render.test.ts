import { isComponentDefinition } from "@juliacomputing/dyad-ast";
import { instantiateModel } from "../../instantiate/model.js";
import { definitionEntity } from "../../workspace/index.js";
import {
  assertHasResult,
  loadModulesFromFS,
} from "../../workspace/utils.test.js";
import { generateIcon } from "../icons/icon.js";
import { renderComponentMarkdown } from "./component.js";
import { documenterOptions } from "./options.js";
import { generateDiagram } from "../icons/diagram.js";
import { flattenDefinitionEntity } from "../../workspace/selectors/flatten.js";
import { getDefinitionNode } from "../../workspace/selectors/nodes.js";

describe("Rendering tests", () => {
  test("Test rendering a single component definition", async () => {
    const { workspace } = await loadModulesFromFS({
      sample: "SampleComponents",
    });

    try {
      const entity = definitionEntity("SampleComponents", [], "Hello");
      const def = assertHasResult(workspace.query(getDefinitionNode(entity)));
      const flat = assertHasResult(
        workspace.query(flattenDefinitionEntity(entity, new Set()))
      );
      expect(isComponentDefinition(def)).toEqual(true);
      expect(isComponentDefinition(flat)).toEqual(true);

      if (isComponentDefinition(def) && isComponentDefinition(flat)) {
        const instance = assertHasResult(
          instantiateModel(def, {}, null, workspace.query.bind(workspace))
        );
        const icon = assertHasResult(
          await generateIcon(
            {},
            def,
            def.name.value,
            0,
            workspace.query.bind(workspace)
          )
        );
        const diagram = assertHasResult(
          await generateDiagram(
            instance.name.value,
            instance,
            workspace.query.bind(workspace)
          )
        );

        const text = renderComponentMarkdown(
          documenterOptions,
          instance,
          def,
          flat,
          [],
          {},
          workspace.query.bind(workspace),
          icon,
          diagram
        );
        expect(text.value).toMatchSnapshot();
      }
    } finally {
      workspace.close();
    }
  });
});
