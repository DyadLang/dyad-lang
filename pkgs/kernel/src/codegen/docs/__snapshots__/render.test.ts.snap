// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering tests Test rendering a single component definition 1`] = `
"\`\`\`@raw html

      <div style="float: right;width: 20%;max-width: 10vw; margin-left: 20px">
      <svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 1000 1000"
    overflow="visible" shape-rendering="geometricPrecision" text-rendering="geometricPrecision">
    <style>
    text {
      transform-origin: center center;
    }
    </style><g transform="translate(0 0) scale(1.0) rotate(0)" transform-origin="center center"><svg xmlns="http://www.w3.org/2000/svg" id="a" width="1000" height="1000" viewBox="0 0 1000 1000"
><defs><style>.b {fill: none;}.b,.c {stroke: #b6b4cf;stroke-linejoin: round;stroke-width: 8px;}.c,.d {  fill: #fff;}.e {fill: #6e6ea1;}.e,.d {stroke-width: 0px;}</style>
  </defs><rect class="d" x="4" y="4" width="992" height="992" /><path class="e" d="M992,8v984H8V8h984M1000,0H0v1000h1000V0h0Z" /><path class="b"
    d="M361.83,226.44c-62.23-3.99-115.91,43.22-119.9,105.45-3.99,62.23,43.22,115.91,105.45,119.9,62.23,3.99,115.91-43.22,119.9-105.45,3.99-62.23-43.22-115.91-105.45-119.9ZM413.76,342.91c-2.09,32.67-30.28,57.46-62.95,55.36-32.67-2.1-57.46-30.28-55.36-62.95,2.09-32.67,30.28-57.46,62.95-55.36,32.67,2.09,57.45,30.28,55.36,62.95Z"
  /><circle class="b" cx="353.85" cy="682.11" r="80.62" /><circle class="b" cx="652.79" cy="644.36" r="52.47" /><polyline class="b" points="410.2 624.82 698.58 340.68 464.78 340.68" /><line class="b" x1="617.63" y1="609.81" x2="432.03" y2="420.13" /><circle class="c" cx="696.98" cy="340.44" r="54.47" />
</svg></g>
    <text x="500" y="1100" transform="rotate(0, 0, 600)" fill="black" font-size="200" font-weight="200" dominant-baseline="central" text-anchor="middle">$(instance)</text>
    </svg>
      <span style="width: 100%; text-align: center; display:inline-block"><code>Hello</code> Icon</span>
      </div>
    
\`\`\`
# \`Hello\`

A simple component model

This component demonstrates several features of the Dyad
modeling language.

## Usage

\`Hello(k=1)\`

## Parameters: 

| Name         | Description                         | Units  |   Default value |
| ------------ | ----------------------------------- | ------ | --------------- |
| \`init\`         |                          | --  |   Initialize.Fixed(x=10) |
| \`k\`         |                          | 1/s  |   1 |



## Variables

| Name         | Description                         | Units  | 
| ------------ | ----------------------------------- | ------ | 
| \`x\`         |                          | m  | 

## Source
\`\`\`dyad
# A simple component model
#
# This component demonstrates several features of the Dyad
# modeling language.
component Hello
  structural parameter init::Initialize = Initialize.Fixed(x=10)
  variable x::Position(statePriority=10)
  parameter k::Real(units="1/s") = 1
  constant g::Acceleration = 9.81
relations
  switch init
    case Steady
      initial der(x) = 0
    end
    case Fixed
      initial x = init.x
    end
  end
  der(x) = -k*x
  assert(x>0 or x<20, "x dropped below 0 or rose above 20")
metadata {
  "Dyad": {
    "experiments": {"world": {"stop": 10}},
    "tests": {"case1": {"stop": 10, "expect": {"initial": {"x": 10}}}}
  }
}
end
\`\`\`

\`\`\`@raw html

<details>
<summary>Flattened Source</summary>

\`\`\`
\`\`\`dyad
# A simple component model
#
# This component demonstrates several features of the Dyad
# modeling language.
component Hello
  structural parameter init::Initialize = Initialize.Fixed(x=10)
  variable x::Position(statePriority=10)
  parameter k::Real(units="1/s") = 1
  constant g::Acceleration = 9.81
relations
  switch init
    case Steady
      initial der(x) = 0
    end
    case Fixed
      initial x = init.x
    end
  end
  der(x) = -k*x
  assert(x>0 or x<20, "x dropped below 0 or rose above 20")
metadata {
  "Dyad": {
    "experiments": {"world": {"stop": 10}},
    "tests": {"case1": {"stop": 10, "expect": {"initial": {"x": 10}}}}
  }
}
end
\`\`\`
\`\`\`@raw html
</details>
\`\`\`
\`\`\`@raw html

<br></br>

\`\`\`

## Related

- Examples
- Experiments
- Analyses"
`;
